Sources/ApolloCodegenLib/Glob.swift

```diff
@@ -0,0 +1,196 @@
+    case invalidExclude(path: String)
+    case unknown(code: Int)
+
+    public var errorDescription: String? {
+      switch self {
+      case .noSpace: return "Malloc call failed" // From Darwin.POSIX.glob
+      case .aborted: return "Unignored error" // From Darwin.POSIX.glob
+      case .cannotEnumerate(let path): return "Cannot enumerate \(path)"
+      case .invalidExclude(let path): return "Exclude paths must start with '!' - \(path)"
+      case .unknown(let code): return "Unknown error: \(code)"
+      }
+    }
+  }
+
+  /// The designated initializer
+  ///
+  /// - Parameters:
+  ///  - pattern: An array of path matching pattern strings.
+  ///
+  /// Each path matching pattern can include the following characters:
+  /// - `*` matches everything but the directory separator (shallow), eg: `*.graphql`
+  /// - `?` matches any single character, eg: `file-?.graphql`
+  /// - `**` matches all subdirectories (deep), eg: `**/*.graphql`
+  /// - `!` excludes any match only if the pattern starts with a `!` character, eg: `!file.graphql`
+  ///
+  /// - Discussion:
+  ///
+  init(_ patterns: [String]) {
+    self.patterns = patterns
+  }
+
+  /// Executes the pattern match on the underlying file system.
+  ///
+  /// - Returns: A set of matched file paths.
+  func match() throws -> Set<String> {
+    let expandedPatterns = try expand(self.patterns)
+
+    var includeMatches: [String] = []
+    var excludeMatches: [String] = []
+
+    for pattern in expandedPatterns {
+      if pattern.isExclude {
+        let patternMatches = try matches(for: String(pattern.dropFirst()))
+        excludeMatches.append(contentsOf: patternMatches)
+
+      } else {
+        let patternMatches = try matches(for: pattern)
+        includeMatches.append(contentsOf: patternMatches)
+      }
+    }
+
+    return Set<String>(includeMatches).subtracting(excludeMatches)
+  }
+
+  /// Separates a comma-delimited string into paths, expanding any globstars and removes duplicates.
+  private func expand(_ patterns: [String]) throws -> Set<String> {
+    var paths: Set<String> = []
+    for pattern in patterns {
+      if pattern.containsExclude && !pattern.isExclude {
+        throw MatchError.invalidExclude(path: pattern)
+      }
+
+      paths.formUnion(pattern.includesGlobstar ? try expandGlobstar(pattern) : [pattern])
+    }
+
+    return paths
+  }
+
+  /// Expands the globstar (`**`) to find all directory paths to search for the match pattern and removes duplicates.
+  private func expandGlobstar(_ pattern: String) throws -> Set<String> {
+    guard pattern.contains("**") else { return [pattern] }
+
+    CodegenLogger.log("Expanding globstar \(pattern)", logLevel: .debug)
+
+    let isExclude = pattern.isExclude
+    var parts = pattern.components(separatedBy: String.Globstar)
+    var firstPart = parts.removeFirst()
+    let lastPart = parts.joined(separator: String.Globstar)
+
+    if isExclude {
+      firstPart = String(firstPart.dropFirst())
```

Comment:
Leaving the `!` in the pattern would cause `glob` to return an empty result. It is added again below so we can continue to track excludes.

Reply:
Document this as well probably.